WalletWasabi/Blockchain/Analysis/BlockchainAnalyzer.cs

```diff
@@ -20,58 +21,61 @@ public BlockchainAnalyzer(int privacyLevelThreshold)
+		var foreignOutputCount = tx.ForeignOutputs.Count;
 
 		if (ownInputCount == 0)
 		{
 			AnalyzeReceive(tx);
 		}
-		else if (inputCount == ownInputCount && outputCount != ownOutputCount)
+		else if (foreignInputCount == 0 && foreignOutputCount > 0)
 		{
 			AnalyzeNormalSpend(tx);
 		}
 		else
 		{
-			AnalyzeWalletInputs(tx, out HashSet<HdPubKey> distinctWalletInputPubKeys, out int newInputAnonset);
+			AnalyzeWalletInputs(tx, out int newInputAnonset, out int sanctionedInputAnonset);
+			AdjustWalletInputsBefore(tx, newInputAnonset);
 
-			if (inputCount == ownInputCount)
+			if (foreignInputCount == 0)
 			{
 				AnalyzeSelfSpend(tx, newInputAnonset);
 			}
 			else
 			{
-				AnalyzeCoinjoin(tx, newInputAnonset, distinctWalletInputPubKeys);
+				AnalyzeCoinjoin(tx, newInputAnonset, sanctionedInputAnonset);
 			}
 
-			AdjustWalletInputs(tx, distinctWalletInputPubKeys, newInputAnonset);
+			AdjustWalletInputsAfter(tx, newInputAnonset);
 		}
 
 		AnalyzeClusters(tx);
 	}
 
 	/// <param name="newInputAnonset">The new anonymity set of the inputs.</param>
-	private void AnalyzeWalletInputs(SmartTransaction tx, out HashSet<HdPubKey> distinctWalletInputPubKeys, out int newInputAnonset)
+	private void AnalyzeWalletInputs(
+		SmartTransaction tx,
+		out int newInputAnonset,
+		out int sanctionedInputAnonset)
 	{
 		// We want to weaken the punishment if the input merge happens in coinjoins.
 		// Our strategy would be is to set the coefficient in proportion to our own inputs compared to the total inputs of the transaction.
 		// However the accuracy can be increased if we consider every input with the same pubkey as a single input entity.
 		// This we can only do for our own inputs as we don't know the pubkeys - nor the scripts - of other inputs.
 		// Another way to think about this is: reusing pubkey on the input side is good, the punishment happened already.
-		distinctWalletInputPubKeys = tx.WalletInputs.Select(x => x.HdPubKey).ToHashSet();
-		var distinctWalletInputPubKeyCount = distinctWalletInputPubKeys.Count;
-		var pubKeyReuseCount = tx.WalletInputs.Count - distinctWalletInputPubKeyCount;
-		double coefficient = (double)distinctWalletInputPubKeyCount / (tx.Transaction.Inputs.Count - pubKeyReuseCount);
+		double coefficient = (double)tx.WalletVirtualInputs.Count / (tx.ForeignInputs.Count + tx.WalletVirtualInputs.Count);
 
-		newInputAnonset = Intersect(distinctWalletInputPubKeys.Select(x => x.AnonymitySet), coefficient);
+		newInputAnonset = Intersect(tx.WalletVirtualInputs.Select(x => x.HdPubKey.AnonymitySet), coefficient);
 
-		foreach (var key in distinctWalletInputPubKeys)
+		CoinjoinAnalyzer coinjoinAnalyzer = new(tx);
+		List<int> anonsets = new();
+		foreach (var virtualInput in tx.WalletVirtualInputs)
 		{
-			key.SetAnonymitySet(newInputAnonset);
+			anonsets.Add(virtualInput.SmartCoins.Select(i => Math.Max(1, virtualInput.HdPubKey.AnonymitySet - (int)coinjoinAnalyzer.ComputeInputSanction(i))).Min());
 		}
```

Comment:
Here I'd also add a comment: Recompute `newInputAnonset` using the discounted anonsets. This result is used as the base for computing the anonymity of this transaction's outputs.